access_group_cache/main.py

```diff
@@ -40,60 +28,137 @@ def _read_secret(name: str) -> Optional[str]:
+            return json.loads(content)['access_token']
+
+
+async def _get_dataset_access_group_members(
+    datasets: List[str],
+) -> Dict[str, List[str]]:
+    access_token = await _get_service_account_access_token()
+
+    group_names = (
+        f'{dataset}-access@populationgenomics.org.au' for dataset in datasets
     )
-
-    members = []
-    page_token = None
-    while True:
-        response = (
-            cloud_identity_service.groups()
-            .memberships()
-            .list(parent=parent, pageToken=page_token)
-            .execute()
+    results = await asyncio.gather(
+        *(
+            _transitive_group_members(access_token, group_name)
+            for group_name in group_names
         )
+    )
 
-        for member in response['memberships']:
-            members.append(member['preferredMemberKey']['id'])
+    return dict(zip(datasets, results))
 
-        page_token = response.get('nextPageToken')
-        if not page_token:
-            break
 
-    all_members = ','.join(members)
+def access_group_cache(unused_data, unused_context):
+    """Cloud Function entry point."""
 
-    # Check whether the current secret version is up-to-date.
-    secret_name = f'{dataset}-access-members-cache'
-    current_secret = _read_secret(secret_name)
+    config = json.loads(_read_secret('server-config'))
 
-    if current_secret == all_members:
-        print(f'Cache for {dataset} is up-to-date.')
-        return  # Nothing left to do.
+    # Google Groups API queries are ridiculously slow, on the order of a few hundred ms
+    # per query. That's why we use async processing here to keep processing times low.
+    group_members = asyncio.run(_get_dataset_access_group_members(config.keys()))
 
-    response = secret_manager.add_secret_version(
-        request={
-            'parent': secret_manager.secret_path(PROJECT_ID, secret_name),
-            'payload': {'data': all_members.encode('UTF-8')},
-        }
-    )
+    for dataset in config:
+        secret_value = ','.join(sorted(group_members[dataset]))
 
-    print(f'Added secret version: {response.name}')
+        # Check whether the current secret version is up-to-date.
+        secret_name = f'{dataset}-access-members-cache'
+        current_secret = _read_secret(secret_name)
 
+        if current_secret == secret_value:
+            print(f'Cache for {dataset} is up-to-date.')
+            continue  # Nothing left to do.
 
-def access_group_cache(unused_data, unused_context):
-    """Main entry point."""
-    config = json.loads(_read_secret('server-config'))
-    # Given that each group takes multiple seconds to process (!), we're running them
-    # concurrently. We're using processes, as google-api-python-client is not thread-safe:
-    # https://googleapis.github.io/google-api-python-client/docs/thread_safety.html
-    with multiprocessing.Pool(20) as pool:
-        pool.map(_process_dataset_group, config)
+        response = secret_manager.add_secret_version(
```

Comment:
Is there a way to disable old versions with this API?

Reply:
Yes, there are actually three different actions you can perform: disable, destroy, and delete: https://googleapis.dev/python/secretmanager/latest/secretmanager_v1/secret_manager_service.html#google.cloud.secretmanager_v1.services.secret_manager_service.SecretManagerServiceAsyncClient.delete_secret

Are you worried that we might accumulate too many versions over time?

Reply:
I have a similar block in a script, and wondering if you can create and disable all in one call (like you can with the UI).

Reply:
I couldn't find anything directly in the API, therefore tried to address this in https://github.com/populationgenomics/cpg-utils/pull/5.

Reply:
I'll merge this for now, as the follow-up PR for using `cpg-utils` will also require using Cloud Run instead of Cloud Functions: a Cloud Function can only depend on pip packages, but we use Anaconda. It'll hopefully be easier to review the respective change by staging the PRs.